src/base/dev/vga/vgaemu.c

```diff
@@ -1145,15 +1145,16 @@ static void vgaemu_update_prot_cache(unsigned page, int prot)
 {
   if(page >= 0xa0 && page < 0xc0) {
     vga.mem.prot_map0[page - 0xa0] = prot;
-    e_invalidate_full(page << PAGE_SHIFT, PAGE_SIZE);
+    /* do not call e_invalidate_full here as that function is not thread-safe */
+    invalidate_unprotected_page_cache(page << PAGE_SHIFT, PAGE_SIZE);
```

Comment:
And why is this one thread-safe?

Reply:
That's a good question. All it does is setting a single dword in the cache to 0xffffffff, which is an atomic operation.

Now if the main thread reads the new 0xffffffff in `mem_likely_protected()` in write_byte() all is fine, since then it will call `vga_write()` which has proper locks.

If the main thread still find the old value and then in between checking and calling WRITE_BYTE, the cache is invalidated, WRITE_BYTE faults because of the mprotect, and the fault handler uses locking too. So I think it's safe but it does mean faults haven't been completely eliminated in sim mode.